src/stateManager/StateManager.ts

```diff
@@ -0,0 +1,778 @@
+        this.p2pEventHooks = p2pEventHooks;
+        this.timeConfig = timeConfig;
+        this.stateChannelManagerContract = stateChannelManagerContract;
+        this.stateChannelEventListener = new StateChannelEventListener(
+            this.self,
+            this.stateChannelManagerContract,
+            this.p2pEventHooks
+        );
+        this.agreementManager = new AgreementManager();
+        this.disputeHandler = new DisputeHandler(
+            this.channelId,
+            signer,
+            signerAddress,
+            this.agreementManager,
+            this.stateChannelManagerContract,
+            this.p2pEventHooks
+        );
+        this.p2pManager = new P2PManager(this.self, signer);
+    }
+    //Mark resources for garbage collection
+    public async dispose() {
+        this.isDisposed = true;
+        this.stateChannelEventListener.dispose();
+        await this.p2pManager.dispose();
+    }
+    public setP2pEventHooks(p2pEventHooks: P2pEventHooks) {
+        this.p2pEventHooks = p2pEventHooks;
+        this.disputeHandler.setP2pEventHooks(p2pEventHooks);
+    }
+    public setChannelId(channelId: BytesLike) {
+        this.channelId = channelId;
+        this.disputeHandler.setChannelId(channelId);
+        this.stateChannelEventListener.setChannelId(channelId);
+    }
+    public getChannelId(): BytesLike {
+        return this.channelId;
+    }
+    public getSignerAddress(): AddressLike {
+        return this.signerAddress;
+    }
+    public getParticipantsCurrent(): Promise<AddressLike[]> {
+        //TODO? this can be done through the AgreementManager for the given fork or thought the stateMachine
+        return this.stateMachine.getParticipants();
+    }
+    public getForkCnt(): number {
+        return this.agreementManager.getLatestForkCnt();
+    }
+    public getNextTransactionCnt(): number {
+        return this.agreementManager.getNextTransactionCnt();
+    }
+    //Triggered by the On-chain Event Listener when a dispute is emitted on-chain
+    public onDisputeUpdate(dispute: DisputeStruct) {
+        this.disputeHandler.onDispute(dispute);
+        this.p2pEventHooks.onDisputeUpdate?.(dispute);
+    }
+    //Triggered by the On-chain Event Listener when block calldata is posted on-chain
+    public collectOnChainBlock(
+        signedBlock: SignedBlockStruct,
+        timestamp: BigNumberish
+    ) {
+        console.log("StateManager - collectOnChainBlock");
+        let flag = this.agreementManager.collectOnChainBlock(
+            signedBlock,
+            Number(timestamp)
+        );
+        let block = EvmUtils.decodeBlock(signedBlock.encodedBlock);
+        let disputeProof: ProofStruct;
+        if (flag == AgreementFlag.DOUBLE_SIGN) {
+            console.log("StateManager - collectOnChainBlock - double sign");
+            disputeProof = this.disputeHandler.createDoubleSignProof([
+                signedBlock
+            ]);
+            this.disputeHandler.createDispute(
```

Comment:
ofcourse here will change as the new structure Is different on creating a dispute

Reply:
This doesn't change much, maybe some refactor - this just detects a specific fraud proof (double sign) and passes it to the dispute handler - the dispute handler will change to incorporate the 'ELVES' logic